### PR TITLE
Add support to compare different numerical types in LOOKUP WHERE clause

### PR DESCRIPTION
**Pull request by Aiee**
1. Refactor the implementation of https://github.com/vesoft-inc/nebula-graph/pull/777 and fix https://github.com/vesoft-inc/nebula-graph/issues/721
2. Add support to LOOKUP queries like ```LOOKUP ON player WHERE player.age >= 40.5 YIELD player.age AS Age```
3. The main idea is to check/cast different numerical values in the validation stage so that optimizer and indexScan rule can keep the current logic.

TODOs:
1. Add the same capability to ```MATCH```
2. extract the typecast logic as a method if we decide to support more cross-type comparisons.

### CODE REVIEW

File: src/validator/LookupValidator.cpp [hunk 1]
```diff
@@ -397,6 +402,24 @@ StatusOr<Value> LookupValidator::checkConstExpr(Expression* expr, const std::str
...
+            auto castVal = v.toInt().first;
+            // edge case col1 >= 40.0, no need to round up
+            if (castVal == v.getFloat()) {
+                return castVal;
+            }
+            return castVal + 1;
+        }
+        return v.toInt().first;
+    }
+
```
Comment (bright-starry-sky) [inline src/validator/LookupValidator.cpp:422]: A small suggestion, It's better to extract a method, such as : 
Value castToTargetType(const Value& v, PropertyType targetType) {
swith (v.getType()) {
case int, float {
 return cast v from current type to target type if needed.
}
case ….. {
return bad value;
}
}
}

File: src/validator/LookupValidator.h [hunk 2]
```diff
@@ -44,7 +44,10 @@ class LookupValidator final : public Validator {
 
     Status rewriteRelExpr(RelationalExpression* expr);
 
-    StatusOr<Value> checkConstExpr(Expression* expr, const std::string& prop);
+    StatusOr<Value> checkConstExpr(Expression* expr,
+                                   const std::string& prop,
+                                   const Expression::Kind kind,
+                                   const bool leftIsAE);
```
Comment (bright-starry-sky) [inline src/validator/LookupValidator.h:50]: Please think about it. Is it necessary to change it here?

**Comment by bright-starry-sky** [comment]
Good job.

File: src/validator/LookupValidator.h (hunk 2 above)
Comment (Aiee) [reply src/validator/LookupValidator.h:50]: To determine whether a value should be cast, it is necessary to get the info about the type of relational expression and also the position of the label attribute.

Comment (bright-starry-sky) [reply src/validator/LookupValidator.h:50]: > To determine whether a value should be cast, it is necessary to get the info about the type of relational expression and also the position of the label attribute.

I think auto `type = schema->getFieldType(prop); `is enough. WDYT?

Comment (Aiee) [reply src/validator/LookupValidator.h:50]: ```schema->getFieldType(prop)``` returns the data type of the column while the parameter```kind```is the kind of relational expression(GT, LT). In this function, ```Expression* expr``` is always a const expression, and I couldn't see another way to determine the comparison operator. Please correct me if I am wrong.

File: src/validator/LookupValidator.cpp (hunk 1 above)
Comment (Aiee) [reply src/validator/LookupValidator.cpp:422]: The reason I didn't extract a function is that the only case we allow different types to compare is when both are numerical, specifically, int and float. I'll modify this part if we decide to support more cross-type comparisons. 
Does it make sense to you?

File: src/validator/LookupValidator.cpp [hunk 3]
```diff
@@ -397,6 +402,27 @@ StatusOr<Value> LookupValidator::checkConstExpr(Expression* expr, const std::str
...
+    } else if (graph::SchemaUtil::propTypeToValueType(type) == Value::Type::INT && v.isFloat()) {
+        // col1 < 10.5 range: [min, 11), col1 < 10 range: [min, 10)
+        // cast float to int and add 1 to get the ceilling
+        if ((leftIsAE && (kind == Expression::Kind::kRelGE || kind == Expression::Kind::kRelLT)) ||
+            (!leftIsAE && (kind == Expression::Kind::kRelGT || kind == Expression::Kind::kRelLE))) {
+            auto castVal = v.toInt();
+            // edge case col1 >= 40.0, no need to round up
+            if (castVal == v.getFloat()) {
+                return castVal;
+            }
```
Comment (yixinglu) [inline src/validator/LookupValidator.cpp:420]:
```cpp
auto f = v.getFloat();
auto i = std::ceil(f);
if (std::abs(f - i) < epsilon) {
  // xxx
}
// yyy
```

File: src/validator/LookupValidator.cpp [hunk 4]
```diff
@@ -387,7 +389,10 @@ Status LookupValidator::rewriteRelExpr(RelationalExpression* expr) {
     return Status::OK();
 }
 
-StatusOr<Value> LookupValidator::checkConstExpr(Expression* expr, const std::string& prop) {
+StatusOr<Value> LookupValidator::checkConstExpr(Expression* expr,
+                                                const std::string& prop,
+                                                const Expression::Kind kind,
+                                                const bool leftIsAE) {
```
Comment (yixinglu) [inline src/validator/LookupValidator.cpp:395]: remove `const` qualifier for primitive parameter.

**Review by bright-starry-sky — APPROVED** [review]
LGTM now， let ours handle other type later.

**Review by yixinglu — APPROVED** [review]

**Pull request merged**